### PR TITLE
Post data-fetch transformation

### LINKED ISSUE #366: Datafetching, serverside post-transformation
## Problem

My API data needs shaping before it's usable on the client. I'd like to be able to have it pretransformed on the server so that I can use it in my preferred shape on the browser

## Solution

In the API editor, at the bottom, add a code editor that allows writing a transformation to be applied to the API result. This should contain a javascript function where the first parameter is the data to be transformed and it returns the transformed object. e.g.

```ts
(data) => Object.entries(data.myProperty).map((data, index) => ({ ...data, id: index }))
```

Apply this transformation serverside before sending the API response

## Note

Make sure to add this to https://github.com/mui/mui-toolpad/issues/365  if that gets implemented first

### PR DESCRIPTION
**Pull request by bharatkashyap**
- Fixes #366 

- [x] I have followed (at least) the [PR section of the contributing guide](https://github.com/mui/material-ui/blob/HEAD/CONTRIBUTING.md#sending-a-pull-request).

### CODE REVIEW

File: packages/toolpad-app/src/server/handleDataRequest.ts [hunk 1]
```diff
@@ -29,6 +29,18 @@ export default async (
   const apiNodeId = req.query.queryId as NodeId;
   const dom = await loadVersionedDom(appId, version);
   const api = appDom.getNode(dom, apiNodeId, 'api');
+  if (api.attributes.transformEnabled.value) {
```
Comment (Janpot) [inline packages/toolpad-app/src/server/handleDataRequest.ts:32]: Won't this fail for existing apis in toolpad? Their `transformEnabled` property and `transform` property will be `undefined`

Comment (bharatkashyap) [reply packages/toolpad-app/src/server/handleDataRequest.ts:32]: Yes, correct - should use optional chaining for the `transform` and `transformEnabled` accesses everywhere

File: packages/toolpad-app/src/server/data.ts [hunk 2]
```diff
@@ -376,7 +377,14 @@ export async function execApi<Q>(
       `Unknown connection "${api.attributes.connectionId.value}" for api "${api.id}"`,
     );
   }
-
+  if (api.attributes.transformEnabled?.value) {
+    const apiResult = await dataSource.exec(connection, api.attributes.query.value, params);
```
Comment (Janpot) [inline packages/toolpad-app/src/server/data.ts:381]: To express the logic more clearly, perhaps extract it out of the `if` statement like:

```tsx
let result  = await exec()
if (transform) {
  result = transform(result)
}
return result;
```

File: packages/toolpad-app/src/server/handleDataRequest.ts [hunk 3]
```diff
@@ -29,6 +29,18 @@ export default async (
...
   const api = appDom.getNode(dom, apiNodeId, 'api');
+  if (api.attributes.transformEnabled?.value) {
+    const apiResult = await execApi(
+      appId,
+      api,
+      req.query.params ? JSON.parse(req.query.params as string) : {},
+    );
+    res.json({
+      data: await evalExpression(
+        `${api.attributes.transform?.value}(${JSON.stringify(apiResult.data)})`,
```
Comment (Janpot) [inline packages/toolpad-app/src/server/handleDataRequest.ts:40]: Isn't it already transformed inside of `execApi`?

Comment (bharatkashyap) [reply packages/toolpad-app/src/server/handleDataRequest.ts:40]: Yes this needn't be there - for some reason I think I found a case in the early exploration when both functions need independent transformation, but testing now and can't find such a case anymore

**Comment by bharatkashyap** [comment]
Two-dot diff for this branch: https://github.com/mui/mui-toolpad/compare/4fa3752..f1c195e

**Review by Janpot — APPROVED** [review]

**Pull request merged**